Modules/_ctypes/_ctypes.c

```diff
@@ -5476,10 +5476,24 @@ comerror_init(PyObject *self, PyObject *args, PyObject *kwds)
     return 0;
 }
 
+static int
+comerror_clear(PyObject *self) {
+    PyBaseExceptionObject *obj = (PyBaseExceptionObject *) self;
+    Py_CLEAR(obj->dict);
+    Py_CLEAR(obj->args);
+    Py_CLEAR(obj->notes);
+    Py_CLEAR(obj->traceback);
+    Py_CLEAR(obj->cause);
+    Py_CLEAR(obj->context);
+    return 0;
+}
+
 static int
 comerror_traverse(PyObject *self, visitproc visit, void *arg)
 {
-    Py_VISIT(Py_TYPE(self));
+    PyTypeObject *tp = Py_TYPE(self);
+    Py_VISIT(tp);
+    tp->tp_clear(self);
```

Comment:
This is not correct; you should not clear anything in the traverse handler.

Reply:
Oh, sorry, I misunderstood your message.
What do you mean here?
> Also, the traverse function must be updated similarly.

Reply:
I _think_ this should work. I'll have a closer look tomorrow.

```suggestion
    tp->tp_traverse(self, visit, arg);
```

Reply:
> Oh, sorry, I misunderstood your message. What do you mean here?
> 
> > Also, the traverse function must be updated similarly.

Sorry, I was unclear. By similarly, I meant that the members we `Py_CLEAR` in the `tp_clear` function, is the same members we `Py_VISIT` in the `tp_traverse` function. Kinda like this:

```c
typedef struct {
    PyObject *a;
    PyObject *b;
    PyObject *c;
} mytype;

static int
mytype_clear(mytype *self)
{
    Py_CLEAR(self->a);
    Py_CLEAR(self->b);
    Py_CLEAR(self->c);
    return 0;
}

static int
mytype_traverse(mytype *self, visitproc visit, void *arg)
{
    // for heap types, always visit self type first
    Py_VISIT(Py_TYPE(self));
    // then, visit the members
    Py_VISIT(self->a);
    Py_VISIT(self->b);
    Py_VISIT(self->c);
    return 0;
}
```

Reply:
Thanks!
If I understood correctly, this will looks like that:
```c
static int
comerror_traverse(PyObject *self, visitproc visit, void *arg)
{
    Py_VISIT(Py_TYPE(self));
    PyBaseExceptionObject *obj = (PyBaseExceptionObject *) self;
    Py_VISIT(obj->dict);
    Py_VISIT(obj->args);
    Py_VISIT(obj->notes);
    Py_VISIT(obj->traceback);
    Py_VISIT(obj->cause);
    Py_VISIT(obj->context);
    return 0;
}
```
But also this touched stuff which belongs to the base class..

Reply:
Not exactly. You should follow this pattern:
https://github.com/python/cpython/blob/080a5961527473af182b025bb29e0c52d43fd49e/Objects/exceptions.c#L1531-L1557

Reply:
`BaseException_clear` is a static function here. Use `PyExc_BaseException->tp_clear` instead.
likewise for the traverse

Reply:
> `BaseException_clear` is a static function here. Use `PyExc_BaseException->tp_clear` instead. likewise for the traverse

Thanks for the answers! =)
Did you mean something like that?
```c
static int
comerror_traverse(PyObject *self, visitproc visit, void *arg)
{
    Py_VISIT(Py_TYPE(self));
    return PyExc_BaseException->tp_clear((PyBaseExceptionObject *) self);
}
```
This also doesn't work. What I'm missing?

Reply:
Why do you need `Py_VISIT(Py_TYPE(self));`?

Reply:
> This also doesn't work.

What does this mean?

Reply:
> > This also doesn't work.
> 
> What does this mean?

Sorry, I was unclear there.
That gives me a compiler errors:
(also, I've translate they to english, because compiler prints error on language of my system... oh, windows..)
```c
C:\Users\KIRILL-1\CLionProjects\cpython\Modules\_ctypes\_ctypes.c(5481,5): warning C4033: "comerror_clear" must return a value [C: 
\Users\KIRILL-1\CLionProjects\cpython\PCbuild\_ctypes.vcxproj]
C:\Users\KIRILL-1\CLionProjects\cpython\Modules\_ctypes\_ctypes.c(5488,5): warning C4033: "comerror_traverse" must return a value  
[C:\Users\KIRILL-1\CLionProjects\cpython\PCbuild\_ctypes.vcxproj]
C:\Users\KIRILL-1\CLionProjects\cpython\Modules\_ctypes\_ctypes.c(5481,33): error C2039: "tp_clear": is not a member of "_object". [C:\Us 
ers\KIRILL-1\CLionProjects\cpython\PCbuild\_ctypes.vcxproj]
C:\Users\KIRILL-1\CLionProjects\cpython\Modules\_ctypes\_ctypes.c(5488,33): error C2039: "tp_traverse": is not a member of "_object". [C: 
\Users\KIRILL-1\CLionProjects\cpython\PCbuild\_ctypes.vcxproj]
    Warnings: 2
    Errors: 2
```

Reply:
I think we are missing `comerror_clear` here.

Reply:
> I think we are missing `comerror_clear` here.

I've wrote this:
```c
static int
comerror_clear(PyObject *self) {
    return PyExc_BaseException->tp_clear((PyBaseExceptionObject *) self);
}

static int
comerror_traverse(PyObject *self, visitproc visit, void *arg)
{
    Py_VISIT(Py_TYPE(self));
    return PyExc_BaseException->tp_traverse((PyBaseExceptionObject *) self);
}
```
Sorry, but I doesn't understand your message. Can you explain?

Reply:
Oh, we need a conversion
```c
    ((PyTypeObject *)PyExc_BaseException)->tp_traverse(...);
```

Reply:
> Oh, we need a conversion
> 
> ```c
>     ((PyTypeObject *)PyExc_BaseException)->tp_traverse(...);
> ```

Finally.. this works! Thank you! I'll commit this. Let's see what Erlend will say.